tools/quantization/quantize.py

```diff
@@ -0,0 +1,899 @@
+
+    @staticmethod
+    def is_integer_ops_mode(mode):
+        return mode in [QuantizationMode.IntegerOps_Static, QuantizationMode.IntegerOps_Dynamic]
+
+    @staticmethod
+    def is_qlinear_ops_mode(mode):
+        return mode in [QuantizationMode.QLinearOps_Static, QuantizationMode.QLinearOps_Dynamic]
+
+    @staticmethod
+    def is_static_mode(mode):
+        return mode in [QuantizationMode.IntegerOps_Static, QuantizationMode.QLinearOps_Static]
+
+# Data Quantization mode
+# Linear_NonScaled: Quantize data using linear, non scaled tranformation.
+# Linear_Scaled: Quantize data using linear, scaled transformation.
+class DataQuantizationMode():
+    Linear_NonScaled = 0
+    Linear_Scaled = 1
+
+quantization_modes = [getattr(QuantizationMode, attr) for attr in dir(QuantizationMode)\
+    if not callable(getattr(QuantizationMode, attr)) and not attr.startswith("__")]
+data_quantization_modes = [getattr(DataQuantizationMode, attr) for attr in dir(DataQuantizationMode)\
+    if not callable(getattr(DataQuantizationMode, attr)) and not attr.startswith("__")]
+
+
+class Weight:
+    '''
+        Represents a linearly quantized weight input from ONNX operators
+    '''
+    def __init__(self, name, initializer, rmins, rmaxs, zero_points, scales, data=[], quantized_data=[], axis=None):
+        self.name = name
+        self.initializer = initializer  # TensorProto initializer in ONNX graph
+        self.rmins = rmins  # List of minimum range for each axis
+        self.rmaxs = rmaxs  # List of maximum range for each axis
+        self.zero_points = zero_points  # 1D tensor of zero points computed for each axis. scalar if axis is empty
+        self.scales = scales  # 1D tensor of scales computed for each axis. scalar if axis is empty
+        self.data = data  # original data from initializer TensorProto
+        self.quantized_data = quantized_data  # weight-packed data from data
+        self.axis = axis  # Scalar to specify which dimension in the initializer to weight pack.
+                          # If empty, single zero point and scales computed from a single rmin and rmax
+
+
+def quantize_data(data, quantize_range, mode=DataQuantizationMode.Linear_NonScaled):
+    '''
+        :parameter quantize_range: list of data to weight pack.
+        :parameter mode: mode to quantize data of type DataQuantizationMode
+        :return: minimum, maximum, zero point, scale, and quantized weights
+
+        To pack weights, we compute a linear transformation
+            - in non-scaled mode, from [rmin, rmax] -> [0, 2^{b-1}] and
+            - in scaled mode, from [-m , m] -> [-(2^{b-1}-1), 2^{b-1}-1] where
+                m = max(abs(rmin), abs(rmax))
+
+        and add necessary intermediate nodes to trasnform quantized weight to full weight using the equation
+        r = S(q-z), where
+            r: real original value
+            q: quantized value
+            S: scale
+            z: zero point
+    '''
+    rmin = min(data)
```

Comment:
what happens when range does not include 0? For example is the range is 2-10 then in this case we dont have a unique representation for 0. Can you do rmin = min(min(data), 0) and similar for max...